File: gui/src/app/lib/filter-dsl/printer.ts
import {BoolFilter, CompFilter, Filter, Visitor} from './filter';
import {Renderer2} from '@angular/core';
import {ValueKind} from './value';

export function PrintPlainText(f: Filter): string {
  const p = new PlainTextPrinter();
  f.accept(p);
  return p.result();
}

export class PlainTextPrinter implements Visitor {

  private buff: string[] = [];

  constructor() {
  }

  result(): string {
    return this.buff.join(' ');
  }

  visitBoolFilter(f: BoolFilter): Visitor {

    const needsLparen = f.lhs.precedence() < f.precedence();
    if (needsLparen) {
      this.buff.push('(');
    }
    f.lhs.accept(this);
    if (needsLparen) {
      this.buff.push(')');
    }
    this.buff.push(f.op);

    const needsRparen = f.rhs.precedence() < f.precedence();
    if (needsRparen) {
      this.buff.push('(');
    }
    f.rhs.accept(this);
    if (needsRparen) {
      this.buff.push(')');
    }
    return this;
  }

  visitCompFilter(f: CompFilter): Visitor {
    if (f.value.kind === ValueKind.String) {
      this.buff.push(f.field, f.op, `"${f.value.v}"`);
    } else if (f.value.kind === ValueKind.Regexp) {
      this.buff.push(f.field, f.op, `/${f.value.v}/`);
    } else {
      this.buff.push(f.field, f.op, '' + f.value.v);
    }
    return this;
  }
}

export function PrintHTML(renderer: Renderer2, f: Filter): HTMLElement {
  const p = new HTMLPrinter(renderer);
  f.accept(p);
  return p.el;
}

export class HTMLPrinter implements Visitor {

  readonly el: HTMLElement;

  constructor(private renderer: Renderer2) {
    this.el = renderer.createElement('span');
  }

  visitBoolFilter(f: BoolFilter): Visitor {

    const boolEl = this.span(['bool-filter'], '');

    const needsLparen = f.lhs.precedence() < f.precedence();

    if (needsLparen) {
      boolEl.appendChild(this.span(['paren'], '('));
    }

    const lhsPrinter = new HTMLPrinter(this.renderer);
    f.lhs.accept(lhsPrinter);
    boolEl.appendChild(lhsPrinter.el);

    if (needsLparen) {
      boolEl.appendChild(this.span(['paren'], ')'));
    }

    boolEl.appendChild(this.span(['bool-op'], ` ${f.op} `));

    if (f.rhs) {
      const needsRparen = f.rhs.precedence() < f.precedence();

      if (needsRparen) {
        boolEl.appendChild(this.span(['paren'], '('));
      }

      const rhsPrinter = new HTMLPrinter(this.renderer);
      f.rhs.accept(rhsPrinter);
      boolEl.appendChild(rhsPrinter.el);

      if (needsRparen) {
        boolEl.appendChild(this.span(['paren'], ')'));
      }
    }

    this.el.appendChild(boolEl);
    return this;
  }

  visitCompFilter(f: CompFilter): Visitor {

    let value: string;
    switch (f.value.kind) {
      case ValueKind.String:
        value = `"${f.value.v}"`;
        break;
      default:
        value = `${f.value.v}`;
    }

    const compEl = this.span(['comp-filter'], '');

    compEl.appendChild(this.span(['field'], `${f.field} `));
    compEl.appendChild(this.span(['comp-op'], `${f.op} `));
    compEl.appendChild(this.span(['value'], value));

    this.el.appendChild(compEl);
    return this;
  }

  private span(cl: string[], innerText: string): HTMLElement {
    const el = this.renderer.createElement('span');
    el.className = [...cl, 'filter-el'].join(' ');
    el.textContent = innerText;
    return el;
  }
}

File: server/pkg/filter/printer.go
package filter

import (
	"bytes"
	"fmt"
	"io"
)

func MustPrint(f Filter) string {
	if f == nil {
		return ""
	}
	s, err := Print(f)
	if err != nil {
		panic(err)
	}
	return s
}

func Print(filter Filter) (string, error) {
	buff := &bytes.Buffer{}
	if err := Fprint(filter, buff); err != nil {
		return "", err
	}
	return buff.String(), nil
}

func Fprint(filter Filter, writer io.Writer) error {
	v := &Printer{w: writer}
	return filter.Accept(v)
}

type Printer struct {
	w io.Writer
}

func (p *Printer) VisitCompFilter(filter *CompFilter) (Visitor, error) {

	if _, err := fmt.Fprintf(p.w, "%s %s %s", filter.Field, filter.Op, filter.Value.String()); err != nil {
		return nil, err
	}
	return p, nil
}

func (p *Printer) VisitBoolFilter(filter *BoolFilter) (Visitor, error) {

	needsLeftParen := filter.LHS.Precedence() < filter.Precedence()

	if needsLeftParen {
		if _, err := fmt.Fprint(p.w, `(`); err != nil {
			return nil, err
		}
	}
	if err := filter.LHS.Accept(p); err != nil {
		return nil, err
	}
	if needsLeftParen {
		if _, err := fmt.Fprint(p.w, `)`); err != nil {
			return nil, err
		}
	}

	if _, err := fmt.Fprintf(p.w, " %s ", string(filter.Op)); err != nil {
		return nil, err
	}

	needsRightParen := filter.RHS.Precedence() < filter.Precedence()

	if needsRightParen {
		if _, err := fmt.Fprint(p.w, `(`); err != nil {
			return nil, err
		}
	}
	if err := filter.RHS.Accept(p); err != nil {
		return nil, err
	}
	if needsRightParen {
		if _, err := fmt.Fprint(p.w, `)`); err != nil {
			return nil, err
		}
	}

	return nil, nil
}
